keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+
+It's not expected that we'll need any configuration "knobs" to allow users to tweak this algorithm: Any users who care about the choice of default IP address (i.e. anyone who is using older servers and clients in their cluster) should simply configure their CNI plugins to only use a single IP family (IPv4-only or IPv6-only). With this configuration, there should only be one pod IP address to choose as the default.
+
+#### PodStatus Internal Representation:
+The PodStatus internal representation will be modified to use a slice of IPs ("PodIPs") rather than a singular IP ("PodIP"):
+```
+    // IP addresses allocated to the pod. Routable at least within the cluster.
+    // Empty if not yet allocated.
+    // +optional
+    PodIPs []string `json:"podIP,omitempty" protobuf:"bytes,6,opt,name=podIP"`
+```
+This "PodIPs" slice representation should eventually become part of a versioned API (after a period of deprecation for the current versioned APIs).
+
+#### kubelet Startup Configuration for Multiple Pod CIDRs
+A new, plural "pod-cidrs" option for the [kubelet startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) is proposed, in addition to retaining the existing, singular "pod-cidr" option (for backwards compatibility):
+```
+  --pod-cidr   string        [Singular IP CIDR,  Default: ""]
+  --pod-cidrs  stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --pod-cidr argument will become deprecated.
+
+#### kube-proxy Startup Configuration for Multiple Pod CIDRs
+A new, plural "cluster-cidrs" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) is proposed, in addition to retaining the existing, singular "cluster-cidr" option (for backwards compatibility):
+```
+  --cluster-cidr   string        [Singular IP CIDR,  Default: ""]
+  --cluster-cidrs  stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --cluster-cidr argument will become deprecated.
+
+### Container Networking Interface (CNI) Plugin Considerations
+
+- This feature requires the use of CNI API 0.3.1 or later.
+- No changes to the CNI API specification (version 0.3.1 or later) are required for Kubernetes dual stack support.
+- A dual-stack capable CNI plugin should return pod IP assignments to kubelet via the "ips" array defined in the [CNI plugin result specification](https://github.com/containernetworking/cni/blob/master/SPEC.md#result).
+- It's expected that plugins will typically not assign more than one IPv4 address per pod.
+- The order of assigned IP addresses that are returned in the CNI plugin result for pod creation will be used to determine the "Default Pod IP" as described in the previous section.
+
+### Support for Multiple Service IPs per Service
+
+This proposal adds the capability to define multiple service CIDRs for a cluster.
```

Comment:
Do we HAVE TO do this?  As a simplifying assumption, what if we said that every cluster has a "primary" address family.  All pods have an address in that family and all Services addresses come from that family.  Pods can have "extra" IPs in other families...

On one hand this seems to simplify Service and Endpoint logic.  On the other hand I am honestly asking "can we get away with this" ?  I don't know.

Or maybe we can get away with this assumption for now, and revisit this later?  The tie-in with multi-net and multi-tenant here is obvious and those topics are not settled, IMO.

Thoughts?

Reply:
Also we should consider how to implement this.  The IP allocation logic depends on etcd, so allocating from a static number of CIDRs should be relatively easy.  Allocating from N will be more complex, and we have the same compat / upgrade / rollback issues to deal with but at a less visible layer of the stack.

Reply:
That is a VERY good question. I've thought about what we would be giving up with this approach. From a high level, I think what we'd end up with is, roughly speaking (assuming an IPv6 service CIDR):
- Internal: V6-only support inside the cluster (no support for legacy IPv4-only pods)
- External: Dual-stack external access to/from the cluster (excluding external load balancers, maybe some ingress controllers?)

More detail (please correct any wrong assumptions that I might have made here):

- IPv4-only pods: IPv4-only client pods couldn't be supported because they could not do service discovery. If these pods do a DNS lookup for a Kubernetes service, they would only care about A records. So we'd have to have an IPv4 address that maps somehow to an IPv6 service IP. We could then potentially do a 4-to-6 mapping of packets, but we'd have to figure out where that mapping gets done (add a sidecar to the pod?). It would be very messy and complicated, at best. Similarly, an IPv4-only backend pod couldn't be supported, unless very complicated DNS address and packet mappings were added.
- Ingress: I believe that this would provide dual-stack access from the outside, even if there in only on service CIDR, at least for an Nginx ingress controller. The Nginx ingress controller should be able to support IPv4 and IPv6 connections from the outside, and load balance them to both IPv4 and IPv6 pod endpoints. (I'm assuming that Kubernetes will still create both IPv4 and IPv6 endpoints, even in the absence of any IPv4 service IPs).
- External Load Balancers: I believe that dual-stack access won't work for at least some external load balancers because they map to a service IP, and let kube-proxy do the load balancing?
- Exporting a Service via NodePort: This could *possibly* provide dual stack support even if there is only a single family service CIDR, BUT... we would have to still support both iptables and ip6tables in parallel, since both tables would need to be listening on that NodePort.
- Exporting a Service via ExternalIP: Same as NodePort, this could *possibly* provide dual stack support even if there is only a single family service CIDR, BUT... we would have to support both iptables and ip6tables in parallel, since iptables would need to be listening for IPv4 ExternalIPs, and ip6tables would need to be listening for IPv6 ExternalIPs.

Reply:
So, I don't have an answer - that's why I was asking. :)

I think all of your points are correct.  It's a little bit of mental gymnastics to think about, but it is simpler overall.  Examples:

* If your "external" LBs support v6, you probably should make v6 your primary (any reason not to?).  Pods can optionally have v4 addresses (why would someone do this?)

* If your "externa;" LBs only support v4, you should make v4 your primary, and some or all pods can have v6 addresses (e.g. intra-corp traffic).

Do we have any real use-cases where someone would want some pods to be dual-stack and some not be?  You have more v6 customers than me :)

Reply:
I'm rewriting this spec to propose using just a single service CIDR (with a section at the end in "Alternatives" discussing the tradeoffs compared to dual service CIDR). I'm thinking that the dual (or more generally, multiple) service CIDR support may take more work than it's worth. I don't know of any specific cases where customers require e.g. an IPv4-only app running in an IPv6-only cluster. As you say, anyone who has IPv4-only (IPv6-only) apps can run their cluster using an IPv4 (IPv6) service CIDR. Plus, in the unlikely event that we find that we *really* need dual service CIDRs, we can add it (or a clever workaround) in a later release.